### PR TITLE
fix: metric deferral + optimizer optimization

### PR DESCRIPTION
**Pull request by Dogacel**
Speedup is near 5-10%, because saved computation is already mostly overlapped. However those optimizations ensure we have more room to overlap in the future.

1. Defer metrics to be reported one step ahead -> ensures main loop is not blocked by `.item()` calls produced by metrics.
2. Replace python for in optimizer with `torch._foreach_copy_` -> %50 faster optimizer steps
3. LRU cache EAGLE-3 block mask (~%5 speedup)

Final profiler result:

image

Before:

image

### CODE REVIEW

File: torchspec/training/optimizer.py [hunk 1]
```diff
@@ -49,6 +49,7 @@ def __init__(
             self.fp32_params,
             lr=lr,
             weight_decay=weight_decay,
+            fused=True,
```
[the remark this answers is not in the document]
Comment (Dogacel) [reply torchspec/training/optimizer.py:52]: > **P2 Badge Avoid forcing fused AdamW for CPU-offloaded training**
> 
> When `fsdp_cpu_offload` is enabled, `apply_fsdp2(..., cpu_offload=True)` installs `CPUOffloadPolicy`, so this optimizer's fp32 master parameters can be CPU tensors. Forcing `fused=True` makes AdamW require fused-supported accelerator tensors and raises on the first optimizer step instead of falling back to the single/foreach implementation, regressing CPU-offloaded training from working to failing; let PyTorch choose the backend or gate this on all master params being on a fused-supported device.
> 
> Useful? React with 👍 / 👎.

Should I disable if fsdp_cpu_offload then?
